File: netDxf/IO/DxfWriter.LightList.cs
using System;
using System.Linq;
using netDxf.Collections;
using netDxf.Objects;

namespace netDxf.IO
{
    internal sealed partial class DxfWriter
    {
        private void PrepareLightListClass(DxfClassCollection definitions)
        {
            bool typed = this.doc.Objects.Items.Any(o => o is DxfLightList);
            int count = this.doc.Objects.Items.Count(o => o.CodeName == "LIGHTLIST");
            if (definitions.Contains("LIGHTLIST"))
            {
                DxfClass definition = definitions["LIGHTLIST"];
                if (definition.CppClassName != "AcDbLightList" || definition.IsEntity)
                {
                    if (typed) throw new System.IO.InvalidDataException("CLASS conflicts with typed LIGHTLIST.");
                    return;
                }
                definition.InstanceCount = count;
            }
            else if (typed) definitions.Add(new DxfClass("LIGHTLIST", "AcDbLightList", "SCENEOE") { ProxyFlags = 1025, IsEntity = false, InstanceCount = count });
        }
        private bool WriteLightListPayload(DxfDatabaseObject item)
        {
            if (!(item is DxfLightList list)) return false;
            this.chunk.Write(100, "AcDbLightList");
            this.chunk.Write(90, list.StoredVersion);
            this.chunk.Write(90, list.Entries.Count);
            foreach (DxfLightListEntry entry in list.Entries)
            {
                this.chunk.Write(5, entry.Light.Handle);
                this.chunk.Write(1, this.EncodeDatabaseString(entry.Name));
            }
            return true;
        }
    }
}

File: javascript/netDxf/IO/DxfWriter.LightList.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfLightList } from '../Objects/DxfLightList.js';import { DxfClass } from '../DxfClass.js';
import { EncodeDxfDatabaseText } from '../../runtime/DxfStringEncoding.js';
import { InvalidDataException } from '../../runtime/Errors.js';
export function PrepareLightListClass(document,definitions){
  const typed=Array.from(document.Objects.Items).some(item=>item instanceof DxfLightList),count=Array.from(document.Objects.Items).filter(item=>item.CodeName==='LIGHTLIST').length;
  if(definitions.Contains('LIGHTLIST')){
    const definition=definitions.get_Item('LIGHTLIST');
    if(definition.CppClassName!=='AcDbLightList'||definition.IsEntity){if(typed)throw new InvalidDataException('CLASS conflicts with typed LIGHTLIST.');return;}definition.InstanceCount=count;
  }else if(typed){const definition=new DxfClass('LIGHTLIST','AcDbLightList','SCENEOE');definition.ProxyFlags=1025;definition.IsEntity=false;definition.InstanceCount=count;definitions.Add(definition);}
}
export function WriteLightListPayload(chunk,version,item){
  if(!(item instanceof DxfLightList))return false;chunk.Write(100,'AcDbLightList');chunk.Write(90,item.StoredVersion);chunk.Write(90,item.Entries.Count);
  for(const entry of item.Entries){chunk.Write(5,entry.Light.Handle);chunk.Write(1,EncodeDxfDatabaseText(entry.Name,version));}return true;
}
